test_plans/spec-constants.asciidoc

```diff
@@ -0,0 +1,562 @@
+  tested type.  The variable's default value has some non-zero value.
+* Declare a single object whose type is a class with `operator(handler &)`.
+  This object will be the command group handler for our test.
+* Create a `queue` from the tested device and call `queue::submit()` twice.
+  Each call passes the same command group handler object described above.
+* Each command group handler sets the value of the spec constant to a
+  different value.
+* Each command group handler submits a kernel via `handler::single_task()`.
+* Each kernel reads the spec constant via
+  `kernel_handler::get_specialization_constant()`.
+* Verify that the value read in the kernel is the same as the value set in
+  the command group handler which launched that kernel instance.
+* The two kernels should run in parallel for this test.
+
+=== Two command groups that read the same spec constant, only one sets value
+
+Same test as in <<sec:two-command>> except only one of the command group
+handlers sets the value of the spec constant.  The kernel instance that is
+launched from the handler that does not set a value should read the spec
+constant's default value.
+
+[[sec:declarations]]
+=== Spec constant defined in various ways
+
+Do the following test for a `specialization_id` variable defined in the
+following ways:
+
+* Defined in a non-global namespace.
+* Defined in the global namespace as `const`.
+* Defined in the global namespace as `constexpr`.
+* Defined in the global namespace as `inline`.
+* Defined in the global namespace as `inline const`.
+* Defined in the global namespace as `inline constexpr`.
+* A static member variable of a struct in the global namespace.
+* A static member variable of a struct in a non-global namespace.
+* A static member variable declared `const` of a struct in the global namespace.
+* A static member variable declared `constexpr` of a struct in the global
+  namespace.
+* A static member variable declared `inline` of a struct in the global
+  namespace.
+* A static member variable declared `inline const` of a struct in the global
+  namespace.
+* A static member variable declared `inline constexpr` of a struct in the global
+  namespace.
+
+The test that is performed is:
+
+* Create a `queue` from the tested device and call `queue::submit()`.
+* Set the value of the spec constant via
+  `handler::set_specialization_constant()`.
+* Submit a kernel via `handler::single_task()`.
+* Read the value of the spec constant via
+  `kernel_handler::get_specialization_constant()` and make sure we get the same
+   value back.
+
+=== Spec constant defined in various ways and set via `kernel_bundle`
+
+* Same test as in <<sec:declarations>>, except:
+  - Set spec constants in a `kernel_bundle`.
+  - Call `build()` to build the `kernel_bundle` into `executable` state.
+  - Register the bundle with a handler via `use_kernel_bundle()`.
+
+[[sec:other-tu]]
+=== Spec constant defined in another translation unit
```

Comment:
What if we have two non-`inline` `specialization_id` defined in different translation units with the same name? This should work and they should be treated as two different specialization constants - do we have this use case listed here somewhere?

Reply:
According to the C++ spec "every program shall contain exactly one definition of every non-inline function or variable that is odr-used in that program outside of a discarded statement (9.4.1); no diagnostic required". Could you possibly clarify the use case?

Reply:
`specialization_id` must be defined as `constexpr` so I assumed it in my comment: `constexpr` implies that the variable will have internal linkage which won't lead to ODR violation.

Reply:
> What if we have two non-inline specialization_id defined in different translation units with the same name? This should work and they should be treated as two different specialization constants - do we have this use case listed here somewhere?

Good idea.  I added this test in 5bbab0683fe2229a1ef27990e5887cf405c68a7f